microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc

```diff
@@ -0,0 +1,109 @@
+= WFLY-12609 MicroProfile JWT 1.1 Support
+:author:            Darran Lofthouse
+:email:             darran.lofthouse@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This proposal is in relation to adding support to WildFly for the MicroProfile JWT 1.1 specification.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12609[WFLY-12609]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1346[EAP7-1346]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+The primary integration will be happening by the addition of a new subsystem `microprofile=jwt` within https://github.com/wildfly/wildfly[WildFly].
+
+Additional integration may be added within https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron] to bridge into the application server's security framework.
+
+=== Other Interested Projects
+
+The integration will also be making use of https://github.com/smallrye/smallrye-jwt[SmallRye JWT] as well as pulling in the https://github.com/eclipse/microprofile-jwt-auth[Eclipse MicroProfile JWT] API library.
+
+== Requirements
+
+=== Hard Requirements
+
+Primarily this enhancement will be delivered using a new subsystem `microprofile-jwt` which will be added to WildFly.  The subsystem will install a DeploymentUnitProcessor which will detect if MicroProfile JWT is required and activate a suitable security configuration as described within this proposal.
+
+At the outset activation will only occur in the following circumstances: -
+
+ * A `@org.eclipse.microprofile.auth.LoginConfig` annotation is detected with an `auth-method` of `MP-JWT`.
+ * An auth method of `MP-JWT` is specified in any web.xml.
+
+This activation scope is deliberately constrained to allow us to incrementally increase the scope at a later point. 
+
+At the point activation occurs the deployment will not make use of a configured `SecurityDomain` instead a "Virtual" `SecurityDomain` will be created to meet the requirements of the deployment and will be used across all sub-deployments.  The common `SecurityDomain` will allow for identity propagation across all sub-deployments whilst the virtual nature of the domain will make it unavailable for propagation across unrelated deployments allowing for deployment level isolation.
```

Comment:
Sub deployments could be in relation to an ear which contains multiple deployments, but it could also be in relation to say a war which contains ejbs where we will also end up deploying across multiple containers.

Reply:
What about case when that ejb from deployment A is calling another ejb from different deployment B? Will be the identity propagated or not ? Deployment A is using MP JWT ...

My guess is that this case falls into "virtual nature of the domain will make it unavailable for propagation across unrelated deployments", but want to be sure.

Reply:
+1 This is one of the main motivations for using a common "virtual security domain" across all sub-deployments - once the identity is established it will be available in all of those deployments.

As a motivation for these deployments is more along the lines of package / deploy / run overall this eliminates the need for complex subsystem configuration for things like identity propagation etc...

I should probably add a statement somewhere but my aim is a developer should be able to take a clean app server and provided the configuration contains the microprofile-jwt subsystem based on our other packaging discussions they should be able to deploy their deployment making use of MicroProfile JWT without being required to make any configuration changes to the server.